File: 序号/733.图像渲染.cpp
class Solution {
  vector<vector<int>> dirs{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

 public:
  /** 与（sr，sc）相同的值且相连，变更为新的color*/
  vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc,
                                int color) {
    int oldColor = image[sr][sc];
    if (oldColor == color) return image;
    deque<pair<int, int>> q;
    int rows = image.size(), cols = image[0].size();
    q.emplace_back(sr, sc);
    int r, c;
    while (!q.empty()) {
      r = q.front().first;
      c = q.front().second;
      q.pop_front();
      image[r][c] = color;
      for (auto& dir : dirs) {
        int nr = r + dir[0], nc = c + dir[1];
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols &&
            image[nr][nc] == oldColor) {
          q.emplace_back(nr, nc);
        }
      }
    }
    return image;
  }
};

File: 序号/733. 图像渲染.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        # dfs
        m,n=len(image),len(image[0])
        oc=image[sr][sc]
        def dfs(r,c):
            if r<0 or r>=m or c<0 or c>=n or image[r][c]!=oc or image[r][c]==color:
                return 
            image[r][c]=color
            dfs(r+1,c)
            dfs(r-1,c)
            dfs(r,c+1)
            dfs(r,c-1)
        dfs(sr,sc)
        return image


        # bfs
        # m,n=len(image),len(image[0])
        # q=deque([(sr,sc)])
        # originColor=image[sr][sc]
        # image[sr][sc]=color
        # directions=[[0,1],[0,-1],[1,0],[-1,0]]
        # while q:
        #     sz=len(q)
        #     for _ in range(sz):
        #         r,c = q.popleft()
        #         print(f'{r} {c}')
        #         for diffX,diffY in directions:
        #             newR,newC=r+diffX,c+diffY
        #             if 0<=newR<m and 0<=newC<n and image[newR][newC]==originColor:
        #                 image[newR][newC]=color
        #                 q.append((newR,newC))
        # return image